### ISSUE TITLE
Missing Village/City API

### ISSUE THREAD

**Issue by ST-DDT** [SpongePowered/SpongeAPI#568]
Recently while playing minecraft survival i stumbled upon a minecraft village.
After a short look in the API i noticed that there is no data object/interface that deal with them in the API.

So i would like to discuss this topic here a bit. It probably has low prio, but IMO there is a lot to discuss.

Here a short list of features, that are/could be connected to the VillageAPI.

- Village/City
  - Create new `Village`s
  - Merge existing `Village`s?
  - Delete existing `Village`s?
  - Add `VillageDoor`s to the `Village`
    - Even doors that otherwise does not meet the requirements of the `Village`?
  - Remove doors from the `Village`
  - Add `Villager`s to the `Village`
  - Remove `Villager`s from a `Village`
  - Change `Village` population ratio (0.35?)
    - Global, Per World or per Village
  - Is the `Village` Stored inside the `Chunk`s or inside the `World`?
  - Is it a plain property/`DataSerializable` or is it something special like `WorldBorder`.
  - Change `Golem` ratios
    - Global, Per World or per Village
  - Change `Village` range (16 blocks?)
    - Global, Per World or per Village
  - Change relation between `Village` and `Player`
- Villager
  - Assign a home a `Villager` (Do they know their home at all? or just their village?)

**Comment by Xemiru** [comment]
i'm pretty sure the Villager AI doesn't decide on a home and just tries to occupy one they randomly select, if not the closest one.

**Comment by gratimax** [comment]
To be honest, something like this does not seem useful at all. All of these effects would be simulated by competent AI code as well as listening to a few events. While there are use cases, they're so few and so specific to Villages themselves that I can't help but think implementing this would be a waste of time.

**Comment by Xemiru** [comment]
what's probably more useful that can branch off to this is structure generation. one can listen for generated villages and handle it accordingly by assigning data that pertains to the generated area; we don't really _need_ a specific api just for villages. if that was the case, we'd need it for every type of generated structure, and that seems a little unclean.

**Comment by ST-DDT** [comment]
@Xemiru The homes for villagers is probably is AI code, you are right.

> if that was the case, we'd need it for every type of generated structure, and that seems a little unclean.

The dungeon for example does not store any data afaik.

@gratimax The most important part for me is adding a door to a village or detect whether a door belongs to a village. I probably don't want to write an entire Villager AI just to add a door to a village.

**Comment by modwizcode** [comment]
@gratimax I disagree on that, I think that the API for supporting Villages is nessicarily specific to villages. Not sure the proposed structure is the best, but if the API matches up with the game itself, this does make sense.

**Comment by Deamon5550** [comment]
At least part of this would be covered by some world generation API pieces I have planned which would add support for structures (villages, mineshafts, nether fortresses, strongholds, desert temples, and those underwater things) after I'm done with the populators.

**Comment by liach** [comment]
Well this is really an issue since the lack of village api affects the implementation of some ais.

**Comment by gabizou** [comment]
> the lack of village api affects the api of some ais.

That's why this is an open issue and not a closed issue.
